### PR TITLE
Earth - Maha

### PR DESCRIPTION
**Pull request by Maha-ElMais**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? | 

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: app/controllers/homepage_controller.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class HomepageController < ApplicationController
+  def index
+    @works = Work.all
+    @movies = Work.all.select { |w| w.category == "movie" }
```
Comment (kaidamasaki) [inline app/controllers/homepage_controller.rb:4]: You should have limited to 10 works sorted by votes.  

For reference you can do that like this:

```suggestion
    @movies = Work.where(category: "movie" ).order(votes: :desc).limit(10)
```

It's probably too late of feedback for you to apply to Video Store API but might be useful to know.

File: app/controllers/votes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+class VotesController < ApplicationController
+  def create
+    @user = User.find_by(id: session[:user_id])
+    @work = Work.find(params[:work_id])
```
Comment (kaidamasaki) [inline app/controllers/votes_controller.rb:4]: It would have been a good idea to check if these return correctly and surface a flash message if they didn't.

File: app/controllers/works_controller.rb [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    @albums = @works.select { |w| w.category == "album" }
+    @books = @works.select { |w| w.category == "book" }
+    @movies = @works.select { |w| w.category == "movie" }
+
+  end
+
+  def top_ten
+    @vote = Vote.find_by(work_id: @work.id)
+  end
```
Comment (kaidamasaki) [inline app/controllers/works_controller.rb:13]: I'm not clear on what this method was for.

File: app/models/vote.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class Vote < ApplicationRecord
+  belongs_to :work
+  belongs_to :user
+
+  validates :user_id, uniqueness: { scope: :work, message: "One user_id per work" }
+
+  def user_votes_table_row
+    work = self.work
+    return [work.title, work.creator, work.publication, work.category.capitalize, self.created_at, work.id]
+  end
```
Comment (kaidamasaki) [inline app/models/vote.rb:10]: This should probably have been in a view helper since it's about layout.

File: app/views/users/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+      <th>Votes</th>
+      <th>Joined</th>
+    </tr>
+    </thead>
+    <tbody>
+    <tr>
+      <td><a href="/users/615">dee</a></td>
+      <td>0</td>
+      <td>May 13, 2020</td>
+    </tr>
```
Comment (kaidamasaki) [inline app/views/users/index.html.erb:17]: This should have pulled from the users table instead of being hardcoded.

File: config/routes.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  root 'homepage#index'
+
+  get '/works/new', to: 'works#new', as: 'new_work'
+  get '/works/:id', to: 'works#show', as: "work"
+  get '/works', to: 'works#index', as: 'works'
+  get  '/works/:id/edit', to: 'works#edit', as: 'edit_work'
+  # put '/works/:id', to: 'works#update'
+  patch '/works/:id', to: 'works#update'
+  post '/works', to: 'works#create'
+  delete '/works/:id', to: 'works#destroy', as: 'destroy_work'
```
Comment (kaidamasaki) [inline config/routes.rb:25]: You can replace these with a single `resources`:

```suggestion
  resources :works
```

**Review by kaidamasaki — COMMENTED** [review]
This was rough, as was expected but thanks for turning it in!  It's behind you now!

# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | More than 10 per list.
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | Page errors.
| 4. Can edit this work successfully, and get a flash message | No flash message.
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | Unsorted.
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | No flash message (silent failure).
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | Shows hardcoded user.
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | Does not change UI.
| 12. Your username is listed in "View all users" page | Only shows hardcoded user.
| 13. Verify that number of votes determines the Media Spotlight | Not implemented.
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | Not implemented.
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | No flash message.
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | Not implemented.
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | No flash message.

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | Not deployed.
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | Styling doesn't match.
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️?

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | Does not use resources.
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | Partials not used.
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | No top ten method.
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | 
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | No fixtures.
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | No model tests.
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | No model tests.
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | No model tests.
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | No model tests.

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | 
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention | 
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |  ✔️
